# Actividad - Proyecto práctico


> La actividad se desarrollará en grupos pre-definidos de 2-3 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.
*   Alumno 1: García Giadás, Ana
*   Alumno 2: Rodriguez Dominguez, Ana
*   Alumno 3: Slonchenko Koberidze, Anton






---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [ ]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/gdrive'
drive_root = mount + "/My Drive/0- Master VIU/Reinforcement Learning-Proyecto"

try:
  from google.colab import drive
  IN_COLAB=True
except:
  IN_COLAB=False

---
### 1.3. Montar carpeta de datos local (solo Colab)

In [ ]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)

    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

We're running Colab
Colab: mounting Google drive on  /content/gdrive
Mounted at /content/gdrive

Colab: making sure  /content/gdrive/My Drive/0- Master VIU/Reinforcement Learning-Proyecto  exists.

Colab: Changing directory to  /content/gdrive/My Drive/0- Master VIU/Reinforcement Learning-Proyecto
/content/gdrive/My Drive/0- Master VIU/Reinforcement Learning-Proyecto
Archivos en el directorio: 
['Anton', 'Ana R', 'Ana G', 'More', 'Borrador.ipynb', 'Proyecto_práctico.ipynb']


---
### 1.4. Instalar librerías necesarias

In [ ]:
if IN_COLAB:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install keras-rl2==1.0.5
  %pip install tensorflow==2.8
else:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install pyglet==1.5.0
  %pip install h5py==3.1.0
  %pip install Pillow==9.5.0
  %pip install keras-rl2==1.0.5
  %pip install Keras==2.2.4
  %pip install tensorflow==2.5.3
  %pip install torch==2.0.1
  %pip install agents==1.4.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 19.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 65.7 MB/s eta 0:00:00
  Created wheel for gym: filename=gym-0.17.3-py3-none-any.whl size=1654622 sha256=547fa759540f7e1d0adca35e47b665658956bf3b7339516cea7ff59c53d96c41
  Stored in directory: /root/.cache/pip/wheels/af/4b/74/fcfc8238472c34d7f96508a63c962ff3ac9485a9a4137afd4e
Successfully built gym
  Attempting uninstall: cloudpickle
    Found existing installation: cloudpickle 2.2.1
    Uninstalling cloudpickle-2.2.1:
      Successfully uninstalled cloudpickle-2.2.1
  Attempting uninstall: gym
    Found existing installation: gym 0.25.2
    Uninstalling gym-0.25.2:
      Successfully uninstalled gym-0.25.2
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning h

---
## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, una solución óptima será alcanzada cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [ ]:
from __future__ import division

from PIL import Image
import numpy as np
import gym

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Flatten, Convolution2D, Permute
from tensorflow.keras.optimizers import Adam
import tensorflow.keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint

#### Configuración base

In [ ]:
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)

np.random.seed(123)
env.seed(123)
nb_actions = env.action_space.n

In [ ]:
class AtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        img = Image.fromarray(observation)
        img = img.resize(INPUT_SHAPE).convert('L')
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')

    def process_state_batch(self, batch):
        processed_batch = batch.astype('float32') / 255.
        return processed_batch

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

1. Implementación de la red neuronal

In [ ]:
input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE # las 4 imagenes que forman un estado se "apilan" en canales
model = Sequential()
print(K.image_data_format())
if K.image_data_format() == 'channels_last':
    # (width, height, channels)
    model.add(Permute((2, 3, 1), input_shape=input_shape))
elif K.image_data_format() == 'channels_first':
    # (channels, width, height)
    model.add(Permute((1, 2, 3), input_shape=input_shape))
else:
    raise RuntimeError('Unknown image_dim_ordering.')

# IMPLEMENTAR ARQUITECTURA AQUI!!!
model.add(Convolution2D(32, (8, 8), strides=(4, 4), activation='relu'))
model.add(Convolution2D(64, (4, 4), strides=(2, 2), activation='relu'))
model.add(Convolution2D(64, (3, 3), activation='relu'))
model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dense(256, activation='relu'))
model.add(Dense(nb_actions, activation='linear'))
model.summary()

channels_last
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 32)        8224      
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_2 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 3136)              0         
                                                                 
 dense (Dense)               (None, 512)               1606144   
                                                                 
 dense_1 (Dense)             (None, 256)  

2. Implementación de la solución DQN

In [ ]:
processor = AtariProcessor()

memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)

policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps',
                                  value_max=1., value_min=.1, value_test=.05,
                                  nb_steps=1000000)

dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy,
                memory=memory, processor=processor,
                nb_steps_warmup=50000, gamma=.99,
                target_model_update=2000,
                train_interval=8)

dqn.compile(Adam(learning_rate=1e-4), metrics=['mae'])

In [ ]:
# Checkpoint
checkpoint_weights_filename = 'dqn_' + env_name + '_weights_{step}.h5f'
callbacks = [ModelIntervalCheckpoint(checkpoint_weights_filename, interval=10000)]
# Log
log_filename = 'dqn_{}_log.json'.format(env_name)
callbacks += [FileLogger(log_filename, interval=10000)]

# Train
dqn.fit(env, callbacks=callbacks, nb_steps=2500000, log_interval=10000, visualize=False)

# Save Weights
weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
dqn.save_weights(weights_filename, overwrite=True)

Training for 2500000 steps ...
Interval 1 (0 steps performed)
   26/10000 [..............................] - ETA: 41s - reward: 0.0000e+00

C:\Anaconda\anaconda3\envs\reinforcement_learning\lib\site-packages\tensorflow\python\keras\engine\training.py:2424: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  warnings.warn('`Model.state_updates` will be removed in a future version. '


10000/10000 [==============================] - 41s 4ms/step - reward: 0.0144
13 episodes - episode_reward: 11.077 [6.000, 19.000] - ale.lives: 2.009

Interval 2 (10000 steps performed)
10000/10000 [==============================] - 41s 4ms/step - reward: 0.0135
16 episodes - episode_reward: 8.438 [4.000, 14.000] - ale.lives: 2.111

Interval 3 (20000 steps performed)
10000/10000 [==============================] - 41s 4ms/step - reward: 0.0111
14 episodes - episode_reward: 7.643 [2.000, 16.000] - ale.lives: 2.153

Interval 4 (30000 steps performed)
10000/10000 [==============================] - 41s 4ms/step - reward: 0.0126
12 episodes - episode_reward: 10.333 [3.000, 17.000] - ale.lives: 2.035

Interval 5 (40000 steps performed)
10000/10000 [==============================] - 41s 4ms/step - reward: 0.0116
14 episodes - episode_reward: 8.571 [2.000, 18.000] - ale.lives: 1.970

Interval 6 (50000 steps performed)
10000/10000 [==============================] - 177s 18ms/step - reward: 0.0135

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



10000/10000 [==============================] - 411s 41ms/step - reward: 0.0186
12 episodes - episode_reward: 16.417 [7.000, 24.000] - loss: 0.017 - mae: 3.360 - mean_q: 4.043 - mean_eps: 0.100 - ale.lives: 2.100

Interval 134 (1330000 steps performed)
10000/10000 [==============================] - 410s 41ms/step - reward: 0.0202
10 episodes - episode_reward: 19.400 [11.000, 29.000] - loss: 0.017 - mae: 3.305 - mean_q: 3.979 - mean_eps: 0.100 - ale.lives: 2.255

Interval 135 (1340000 steps performed)
10000/10000 [==============================] - 410s 41ms/step - reward: 0.0195
12 episodes - episode_reward: 16.667 [9.000, 27.000] - loss: 0.017 - mae: 3.305 - mean_q: 3.979 - mean_eps: 0.100 - ale.lives: 2.165

Interval 136 (1350000 steps performed)
10000/10000 [==============================] - 410s 41ms/step - reward: 0.0196
12 episodes - episode_reward: 17.167 [10.000, 24.000] - loss: 0.016 - mae: 3.305 - mean_q: 3.977 - mean_eps: 0.100 - ale.lives: 2.138

Interval 137 (1360000 steps p

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



10000/10000 [==============================] - 419s 42ms/step - reward: 0.0225
12 episodes - episode_reward: 19.333 [9.000, 35.000] - loss: 0.018 - mae: 3.377 - mean_q: 4.066 - mean_eps: 0.100 - ale.lives: 2.081

Interval 178 (1770000 steps performed)
10000/10000 [==============================] - 410s 41ms/step - reward: 0.0216
14 episodes - episode_reward: 15.500 [7.000, 23.000] - loss: 0.017 - mae: 3.294 - mean_q: 3.966 - mean_eps: 0.100 - ale.lives: 2.054

Interval 179 (1780000 steps performed)
10000/10000 [==============================] - 410s 41ms/step - reward: 0.0214
12 episodes - episode_reward: 17.250 [9.000, 30.000] - loss: 0.018 - mae: 3.270 - mean_q: 3.939 - mean_eps: 0.100 - ale.lives: 2.219

Interval 180 (1790000 steps performed)
10000/10000 [==============================] - 410s 41ms/step - reward: 0.0204
14 episodes - episode_reward: 15.071 [7.000, 21.000] - loss: 0.018 - mae: 3.276 - mean_q: 3.947 - mean_eps: 0.100 - ale.lives: 2.271

Interval 181 (1800000 steps per

In [ ]:
# Testing part to calculate the mean reward
weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)

Testing for 10 episodes ...
Episode 1: reward: 25.000, steps: 966
Episode 2: reward: 26.000, steps: 998
Episode 3: reward: 25.000, steps: 917
Episode 4: reward: 23.000, steps: 913
Episode 5: reward: 21.000, steps: 876
Episode 6: reward: 34.000, steps: 1218
Episode 7: reward: 16.000, steps: 763
Episode 8: reward: 30.000, steps: 1172
Episode 9: reward: 9.000, steps: 526
Episode 10: reward: 23.000, steps: 840


In [ ]:
# Calculamos la media de los resultados anteriores
np.mean(dqn.history.history['episode_reward'])

23.2

Para confirmar que estos resultados eran fiables, hemos lanzado un Notebook aparte cargando los pesos y ejecutando test 5 veces consecutivas. Todos los resultados medios fueron superiores a 20 puntos, consiguiendo incluso mejorar la puntuación máxima hasta 26.4 puntos:

In [ ]:
weights_filename = "dqn_SpaceInvaders-v0_weights.h5f"
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
print("\nRecompensa media:",np.mean(dqn.history.history['episode_reward']))

Testing for 10 episodes ...
Episode 1: reward: 22.000, steps: 714
Episode 2: reward: 28.000, steps: 1060
Episode 3: reward: 35.000, steps: 1468
Episode 4: reward: 22.000, steps: 1086
Episode 5: reward: 27.000, steps: 1136
Episode 6: reward: 27.000, steps: 833
Episode 7: reward: 20.000, steps: 838
Episode 8: reward: 32.000, steps: 1316
Episode 9: reward: 32.000, steps: 1351
Episode 10: reward: 19.000, steps: 680

Recompensa media: 26.4


In [ ]:
weights_filename = "dqn_SpaceInvaders-v0_weights.h5f"
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
print("\nRecompensa media:",np.mean(dqn.history.history['episode_reward']))

Testing for 10 episodes ...


/usr/local/lib/python3.10/dist-packages/keras/engine/training_v1.py:2079: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 11.000, steps: 495
Episode 2: reward: 22.000, steps: 837
Episode 3: reward: 27.000, steps: 1069
Episode 4: reward: 22.000, steps: 1000
Episode 5: reward: 15.000, steps: 625
Episode 6: reward: 25.000, steps: 970
Episode 7: reward: 8.000, steps: 399
Episode 8: reward: 22.000, steps: 774
Episode 9: reward: 32.000, steps: 1436
Episode 10: reward: 30.000, steps: 1044

Recompensa media: 21.4


In [ ]:
weights_filename = "dqn_SpaceInvaders-v0_weights.h5f"
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
print("\nRecompensa media:",np.mean(dqn.history.history['episode_reward']))

Testing for 10 episodes ...
Episode 1: reward: 26.000, steps: 975
Episode 2: reward: 13.000, steps: 508
Episode 3: reward: 33.000, steps: 1202
Episode 4: reward: 19.000, steps: 662
Episode 5: reward: 15.000, steps: 671
Episode 6: reward: 6.000, steps: 352
Episode 7: reward: 31.000, steps: 1158
Episode 8: reward: 28.000, steps: 1153
Episode 9: reward: 34.000, steps: 1097
Episode 10: reward: 22.000, steps: 1012

Recompensa media: 22.7


In [ ]:
weights_filename = "dqn_SpaceInvaders-v0_weights.h5f"
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
print("\nRecompensa media:",np.mean(dqn.history.history['episode_reward']))

Testing for 10 episodes ...
Episode 1: reward: 21.000, steps: 704
Episode 2: reward: 16.000, steps: 691
Episode 3: reward: 13.000, steps: 618
Episode 4: reward: 20.000, steps: 856
Episode 5: reward: 27.000, steps: 963
Episode 6: reward: 25.000, steps: 1036
Episode 7: reward: 20.000, steps: 717
Episode 8: reward: 29.000, steps: 1223
Episode 9: reward: 20.000, steps: 579
Episode 10: reward: 21.000, steps: 883

Recompensa media: 21.2


In [ ]:
weights_filename = "dqn_SpaceInvaders-v0_weights.h5f"
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
print("\nRecompensa media:",np.mean(dqn.history.history['episode_reward']))

Testing for 10 episodes ...
Episode 1: reward: 18.000, steps: 681
Episode 2: reward: 19.000, steps: 828
Episode 3: reward: 26.000, steps: 904
Episode 4: reward: 27.000, steps: 1059
Episode 5: reward: 24.000, steps: 984
Episode 6: reward: 16.000, steps: 656
Episode 7: reward: 22.000, steps: 750
Episode 8: reward: 32.000, steps: 1199
Episode 9: reward: 16.000, steps: 642
Episode 10: reward: 20.000, steps: 776

Recompensa media: 22.0


3. Justificación de los parámetros seleccionados y de los resultados obtenidos

Hemos probado 4 arquitecturas diferentes subiendo de complejidad. El modelo que obtuvo los mejores resultados es el que se presentó en los apartados anteriores, consiguiendo un **resultado en test de 23.2** (mejorado hasta 26.4 en pruebas sucesivas).

Los hiperparámetros se fueron seleccionando en base a diferentes pruebas y observaciones, resultando en los siguientes cambios en comparación con lo visto en clase:
- target_model_update=2000
- train_interval=8
- learning_rate=1e-4
- entrenamiento durante 2.5 millones de iteraciones.

El hiperparámetro "target_model_update" se estableció en 2000, lo que implicó actualizaciones regulares del modelo objetivo. Esta configuración permitió mantener una estimación más precisa del valor de las acciones y promovió la estabilidad del aprendizaje alineando el modelo objetivo con el modelo principal.

El hiperparámetro "train_interval" se fijó en 8, lo que implicó actualizaciones frecuentes del modelo durante el entrenamiento. Esta configuración permitió una adaptación más ágil a los cambios del entorno y mejoró la eficiencia del proceso de entrenamiento.

El hiperparámetro "learning_rate" se estableció en 1e-4. Al seleccionar un valor bajo, se permitió un aprendizaje más lento y gradual, evitando cambios bruscos que podrían afectar negativamente al rendimiento del agente. Esto permitió una convergencia más estable y la capacidad de aprender patrones sutiles en el entorno.

Para facilitar la exploración durante el entrenamiento hemos utilizado el algoritmo LinearAnnealedPolicy. Esto nos permitió ir reduciendo el valor de epsilon (probabilidad de tomar acción aleatoria) desde el valor máximo del 100% hasta el 10%, de manera lineal y a lo largo de 1 millón de pasos. Además se estableció un 5% de probabilidad de acción aleatoria durante el test.

Se realizó el entrenamiento durante 2.5 millones de iteraciones, lo que proporcionó al modelo suficiente tiempo para adaptarse y aprender de manera continua sin interrupciones. Este modelo fue el único que hemos conseguido entrenar durante tantos pasos sin ninguna interrupción, lo que en nuestra opinión fue un factor importante para conseguir los mejores resultados.

A continuación describimos las dificultades con las que nos hemos encontrado durante la realización del presente proyecto, así como las diferentes arquitecturas probadas y su justificación.

**DIFICULTADES**

Los primeros entrenamientos fueron realizados en Google Colab porque era más sencillo y rápido de configurar y porque había integrantes del grupo cuyas máquinas no admitían configuración en el entorno local. En Colab, la principal parte de los entrenamientos fue realizada en entornos CPU, aunque vimos que modelos pesados se entrenaban considerablemente más rápido en GPU. El problema principal fueron interrupciones frecuentes durante entrenamiento en Colab debido a diferentes causas: captcha "No soy robot", desconexión temprana por el límite de uso de GPU, u otras razones que no se podían identificar a la mañana siguiente. El problema persistió incluso tras adquirir unidades de computación a través de servicio de pago.

Hemos observado que en todos los casos cuando se presentaban interrupciones, los resultados tanto durante el entrenamiento como luego en test iban mejorando hasta un cierto momento, tras el cual empezaban a empeorar hasta el punto que un modelo entrenado durante 2 millones de pasos, pero con interrupciones, daba peores resultados que un modelo sin entrenar.

Por supuesto en todo momento hemos utilizado los checkpoints, y hemos automatizado el re-calculo de los parametros de epsilon, cuyo código se presenta en el anexo. Hemos pensado que se podía provocar por el hecho de que cada nuevo entrenamiento comienza con la memoria vacía, y para solucionarlo hemos probado grabar la memoria en un fichero con la libreria pickle. No obstante no pudimos aplicar este metodo, debido a que pickle guarda sin comprimir, por lo que el fichero se vuelve pesado rápidamente, saturando la RAM del Colab.

Finalmente hemos pasado a entorno local, donde tras solucionar también una serie de dificultades, hemos conseguido entrenamientos continuos que finalmente presentaron los mejores resultados.

**ARQUITECTURAS PROBADAS**

Hemos comenzado nuestras pruebas con la implementación de la arquitectura presentada en el artículo Playing Atari with Deep Reinforcement Learning (https://arxiv.org/abs/1312.5602). Desafortunadamente trás 2 millones de steps los resultados en test apenas superaban los de un modelo sin entrenar. Creemos que se debe a las interrupciones en el entrenamiento, quizás un entrenamiento continuo daría mejores resultados. Cabe mencionar que los autores del artículo entrenaron sus modelos durante 10 millones de pasos.

In [ ]:
model.add(Convolution2D(16, (8, 8), strides=(4, 4)))
model.add(Activation('relu'))
model.add(Convolution2D(32, (4, 4), strides=(2, 2)))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(256))
model.add(Activation('relu'))
model.add(Dense(nb_actions))
model.add(Activation('linear'))
print(model.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 16)        4112      
                                                                 
 activation (Activation)     (None, 20, 20, 16)        0         
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 32)          8224      
                                                                 
 activation_1 (Activation)   (None, 9, 9, 32)          0         
                                                                 
 flatten (Flatten)           (None, 2592)              0         
                                                                 
 dense (Dense)               (None, 256)               6

En vistas de no conseguir los resultados necesarios, fuimos subiendo la complejidad de nuestro modelo, añadiendo capas convolucionales/densas y subiendo número de filtros/neuronas. En comparación con la arquitectura del articulo, la siguiente red presenta una capa convolucional adicional, subiendo además tanto el número de filtros como el número de neuronas. Con esta red tras 1 millon de steps hemos conseguido un resultado en test de 20.9 , aunque los resultados empeoraron al continuar entrenamiento (con interrupciones) hasta 2 millones de pasos.

In [ ]:
model.add(Convolution2D(32, (8, 8), strides=(4, 4)))
model.add(Activation('relu'))
model.add(Convolution2D(64, (4, 4), strides=(2, 2)))
model.add(Activation('relu'))
model.add(Convolution2D(64, (3, 3), strides=(1, 1)))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(nb_actions))
model.add(Activation('linear'))
print(model.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 32)        8224      
                                                                 
 activation (Activation)     (None, 20, 20, 32)        0         
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 activation_1 (Activation)   (None, 9, 9, 64)          0         
                                                                 
 conv2d_2 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 activation_2 (Activation)   (None, 7, 7, 64)          0

In [ ]:
weights_filename = 'dqn_{}_weights_1080000.h5f'.format(env_name)
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
np.mean(dqn.history.history['episode_reward'])

Testing for 10 episodes ...
Episode 1: reward: 17.000, steps: 786
Episode 2: reward: 16.000, steps: 659
Episode 3: reward: 27.000, steps: 942
Episode 4: reward: 19.000, steps: 1089
Episode 5: reward: 23.000, steps: 1117
Episode 6: reward: 18.000, steps: 865
Episode 7: reward: 24.000, steps: 892
Episode 8: reward: 18.000, steps: 795
Episode 9: reward: 20.000, steps: 824
Episode 10: reward: 27.000, steps: 1226


20.9

In [ ]:
weights_filename = 'dqn_{}_weights_1080000.h5f'.format(env_name)
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
np.mean(dqn.history.history['episode_reward'])

Testing for 10 episodes ...
Episode 1: reward: 24.000, steps: 1036
Episode 2: reward: 16.000, steps: 745
Episode 3: reward: 21.000, steps: 805
Episode 4: reward: 20.000, steps: 1124
Episode 5: reward: 39.000, steps: 1508
Episode 6: reward: 12.000, steps: 673
Episode 7: reward: 11.000, steps: 521
Episode 8: reward: 23.000, steps: 949
Episode 9: reward: 16.000, steps: 863
Episode 10: reward: 27.000, steps: 1035


20.9

A continuación hemos añadido una capa densa adicional con 256 neuronas. Este es el modelo que hemos presentado en la sección del Desarrollo del presente trabajo. Tras un entrenamiento continuo durante 2.5 millones de iteraciones, esta arquitectura fue la que dio los mejores resultados en test de 23.2 puntos de recompensa media.

In [ ]:
model.add(Convolution2D(32, (8, 8), strides=(4, 4), activation='relu'))
model.add(Convolution2D(64, (4, 4), strides=(2, 2), activation='relu'))
model.add(Convolution2D(64, (3, 3), activation='relu'))
model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dense(256, activation='relu'))
model.add(Dense(nb_actions, activation='linear'))
model.summary()

channels_last
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 32)        8224      
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_2 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 3136)              0         
                                                                 
 dense (Dense)               (None, 512)               1606144   
                                                                 
 dense_1 (Dense)             (None, 256)  

La arquitectura final que hemos probado fue la más pesada de todas, incrementando el número de neuronas de las capas densas del modelo anterior a 1024 y 512 respectivamente.

In [ ]:
model.add(Convolution2D(32, (8, 8), strides=(4, 4)))
model.add(Activation('relu'))
model.add(Convolution2D(64, (4, 4), strides=(2, 2)))
model.add(Activation('relu'))
model.add(Convolution2D(128, (3, 3), strides=(1, 1)))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(1024))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(nb_actions))
model.add(Activation('linear'))
print(model.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 32)        8224      
                                                                 
 activation (Activation)     (None, 20, 20, 32)        0         
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 activation_1 (Activation)   (None, 9, 9, 64)          0         
                                                                 
 conv2d_2 (Conv2D)           (None, 7, 7, 128)         73856     
                                                                 
 activation_2 (Activation)   (None, 7, 7, 128)         0

El entrenamiento de la arquitectura mostrada anteriormente se tuvo que realizar de forma fraccionada debido a interrupciones del kernel, consto de 4 entrenamientos, cada vez se interrupía el entrenamiento iniciabamos otro de forma sucesiva utilizando los últimos pesos actualizados del anterior,  pese a las interrupciones mencionadas, cabe descatacar que se entreno en unas 16 horas aproximadamente con un nb_step=2000000 (tiempo bastante razonable para la cantidad de parámetros entrenables disponibles), aun así los resultados que obtuvimos fueron inferiores a los de la arquitectura del modelo que presentamos como candidata a resolver el desafío.

A continuación se muestra la evaluacion en test para 10 episodios:

In [ ]:
# Cargar los últimos pesos
weights_filename = "./Checkpoints/"+ 'dqn_{}_weights_{}.h5f'.format(env_name, last_checkpoint)
dqn.load_weights(weights_filename)
dqn.test(env, nb_episodes=10, visualize=False)
print("\nRecompensa media:",np.mean(dqn.history.history['episode_reward']))

Testing for 10 episodes ...
Episode 1: reward: 15.000, steps: 627
Episode 2: reward: 21.000, steps: 757
Episode 3: reward: 22.000, steps: 849
Episode 4: reward: 22.000, steps: 875
Episode 5: reward: 10.000, steps: 636
Episode 6: reward: 28.000, steps: 1475
Episode 7: reward: 33.000, steps: 1482
Episode 8: reward: 22.000, steps: 888
Episode 9: reward: 16.000, steps: 627
Episode 10: reward: 24.000, steps: 1019

Recompensa media: 21.3


Observamos que las recompensas son bastante inestables en función del episodio que se evalúa, obtenemos recompensas muy buenas de hasta 33 y recompensas de 10, muy por debajo del objetivo marcado.

**ANEXO**

Aquí se presenta el código que fuimos implementando para intentar solucionar el problema de las interrupciones de entrenamiento. Tal y como explicamos anteriormente, los mejores resultados conseguidos finalmente vienen de los entrenamientos sin interrupciones.

In [ ]:
import pickle

def save_memory(memory, filename):
    with open(filename, 'wb') as file:
        pickle.dump(memory, file)

def save_policy_params(attr, value_max, value_min, value_test, nb_steps, filename):
    policy_params = {
    'attr': attr,
    'value_max': value_max,
    'value_min': value_min,
    'value_test': value_test,
    'nb_steps': nb_steps
    }
    with open(filename, 'wb') as file:
        pickle.dump(policy_params, file)

def save_training_info(nb_steps, filename):
    training_info = {
    'nb_steps': nb_steps
    }
    with open(filename, 'wb') as f:
        pickle.dump(training_info, f)

def load_memory(filename):
    with open(filename, 'rb') as f:
        memory = pickle.load(f)
    return memory

def load_policy_params(filename):
    with open(filename, 'rb') as f:
        policy_params = pickle.load(f)
    return policy_params

def load_training_info(filename):
    with open(filename, 'rb') as f:
        training_info = pickle.load(f)
    return training_info


# Variable PRIMER_ENTRENAMIENTO indica:
    #True -> se va a hacer el entrenamiento desde cero.
    #False -> se va a retomar el entrenamiento desde la info almacenada.
try:
    archivos = []
    for archivo in os.listdir("./Checkpoints/Entrenamiento CNN_memoria_2/Weights/"):
        nombre, extension = os.path.splitext(archivo)
        if 'weights_' in nombre:
            num_pasos = nombre.split('_')[3].split('.')[0]
            if '+' in num_pasos:
                num_pasos_act, num_pasos_ant  = num_pasos.split('+')
                num_pasos = int(num_pasos_ant) + int(num_pasos_act)
                archivos.append(num_pasos)
    last_checkpoint = max(archivos)
    PRIMER_ENTRENAMIENTO = False

except:
    last_checkpoint = 0
    PRIMER_ENTRENAMIENTO = True

# Directorio para guardar los pesos y registros
checkpoint_dir = "./Checkpoints/Entrenamiento CNN_memoria_2/Weights/"
os.makedirs(checkpoint_dir, exist_ok=True)
if not PRIMER_ENTRENAMIENTO:
    checkpoint_weights_filename = checkpoint_dir + 'dqn_' + env_name + '_weights_{step}' + '+{}.h5f'.format(last_checkpoint)
else:
    checkpoint_weights_filename = checkpoint_dir + 'dqn_' + env_name + '_weights_{step}.h5f'
print(checkpoint_weights_filename)
weights_filename = checkpoint_dir + 'dqn_{}_weights.h5f'.format(env_name)
log_filename = checkpoint_dir + 'dqn_{}_log.json'.format(env_name)

#Directorio para guardar la memoria
memory_dir = "./Checkpoints/Entrenamiento CNN_memoria_2/Memory/"
os.makedirs(memory_dir, exist_ok=True)
memory_filename = memory_dir + 'dqn_{}_memory.pkl'.format(env_name)

# Directorio para guardar la policy
policy_dir = "./Checkpoints/Entrenamiento CNN_memoria_2/Policy/"
os.makedirs(policy_dir, exist_ok=True)
policy_filename = policy_dir + 'dqn_{}_policy.pkl'.format(env_name)

#Directorio para guardar la info de training
#training_info_dir = "./Checkpoints/Entrenamiento CNN_memoria/Training_info/"
#os.makedirs(training_info_dir, exist_ok=True)
#training_info_filename = training_info_dir + 'dqn_{}_training_info.pkl'.format(env_name)

processor = AtariProcessor()

if PRIMER_ENTRENAMIENTO:
    warmup = 50000

    memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)

    policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps',
                                      value_max=1., value_min=.1, value_test=.2,
                                      nb_steps=1000000)

    dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy,
                    memory=memory, processor=processor,
                    nb_steps_warmup=warmup, gamma=.99,
                    target_model_update=2000,
                    train_interval=8)
    dqn.compile(Adam(learning_rate=1e-4), metrics=['mae'])

else:
    warmup = 0

    memory = load_memory(memory_filename)

    #training_info = load_training_info(training_info_filename)

    policy_params = load_policy_params(policy_filename)

    policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), **policy_params)

    dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy,
                    memory=memory, processor=processor,
                    nb_steps_warmup=warmup, gamma=.99,
                    target_model_update=2000,
                    train_interval=8)
    dqn.compile(Adam(learning_rate=1e-4), metrics=['mae'])

    # Cargar los valores y pesos
    dqn.step = training_info['nb_steps']
    dqn.load_weights(weights_filename)


# Checkpoint
callbacks = [ModelIntervalCheckpoint(checkpoint_weights_filename, interval=10000)]
# Log
callbacks += [FileLogger(log_filename, interval=10000)]

total_steps = 2500000

# Train
if PRIMER_ENTRENAMIENTO:
    dqn.fit(env, callbacks=callbacks, nb_steps=total_steps, log_interval=10000, visualize=False)

else:
    remaining_steps = total_steps - last_checkpoint
    dqn.fit(env, callbacks=callbacks, nb_steps=remaining_steps, log_interval=10000, visualize=False)

# Save Weights
dqn.save_weights(weights_filename, overwrite=True)
# Save memory
save_memory(dqn.memory, memory_filename)
# Save policy
save_policy_params(dqn.policy.attr, dqn.policy.value_max, dqn.policy.value_min, dqn.policy.value_test, dqn.policy.nb_steps, policy_filename)
#Save training info
save_training_info(dqn.step, training_info_filename)

---